File: scripts/check-sensitive-content.ts
import { readdir } from 'node:fs/promises';
import { join, relative } from 'node:path';

const repositoryRoot = new URL('../', import.meta.url).pathname;
const excludedDirectories = new Set(['.git', 'dist', 'node_modules']);
const excludedFiles = new Set(['bun.lock']);
const patterns = [
  /-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----/,
  /\bgh[opsu]_[A-Za-z0-9]{20,}\b/,
  /\bAKIA[A-Z0-9]{16}\b/,
  /\bxox[baprs]-[A-Za-z0-9-]{20,}\b/,
  /\b(?:sk|rk)_(?:live|test)_[A-Za-z0-9]{20,}\b/,
  /\bfspat_[A-Za-z0-9_-]{32,}\b/,
  /\bfsinv_[A-Za-z0-9_-]{32,}\b/,
  /\b(?:DATABASE_URL|POSTGRES_PASSWORD|S3_SECRET_ACCESS_KEY|DOKPLOY_API_KEY)\s*=/i,
  /https?:\/\/[^\s/:]+:[^\s/@]+@[^\s/]+/,
];

const findings: string[] = [];

async function scanDirectory(directory: string): Promise<void> {
  for (const entry of await readdir(directory, { withFileTypes: true })) {
    if (entry.isDirectory() && excludedDirectories.has(entry.name)) continue;
    if (entry.isFile() && excludedFiles.has(entry.name)) continue;
    const path = join(directory, entry.name);
    if (entry.isDirectory()) {
      await scanDirectory(path);
      continue;
    }
    if (!entry.isFile()) continue;
    const content = await Bun.file(path).text();
    for (const [index, line] of content.split('\n').entries()) {
      if (patterns.some((pattern) => pattern.test(line))) {
        findings.push(`${relative(repositoryRoot, path)}:${index + 1}`);
      }
    }
  }
}

await scanDirectory(repositoryRoot);
if (findings.length > 0) {
  throw new Error(`Potential sensitive content found at ${findings.join(', ')}`);
}
console.error('No sensitive credential material detected in the public source tree.');

File: scripts/check_sensitive_content.py
#!/usr/bin/env python3
"""Reject credentials and browser sign-in artifacts in the public plugin source tree."""

from __future__ import annotations

import re
from pathlib import Path


ROOT = Path(__file__).resolve().parent.parent
PATTERNS = (
    ("private key", re.compile(r"-----BEGIN (?:RSA |EC |OPENSSH )?PRIVATE KEY-----")),
    ("GitHub token", re.compile(r"\bgh[opsu]_[A-Za-z0-9]{20,}\b")),
    ("AWS access key", re.compile(r"\bAKIA[A-Z0-9]{16}\b")),
    ("Slack token", re.compile(r"\bxox[baprs]-[A-Za-z0-9-]{20,}\b")),
    ("payment secret", re.compile(r"\b(?:sk|rk)_(?:live|test)_[A-Za-z0-9]{20,}\b")),
    ("FeedbackServer legacy token", re.compile(r"\bfs(?:at|rt)_[A-Za-z0-9_-]{32,}\b")),
    ("FeedbackServer personal access token", re.compile(r"\bfspat_[A-Za-z0-9_-]{24,}\b")),
    ("bearer JWT", re.compile(r"\bBearer\s+eyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\b", re.I)),
    ("browser session", re.compile(r"\b(?:better-auth\.)?session_token\s*=\s*[^\s;]{16,}", re.I)),
    ("live magic link", re.compile(r"/magic-link/verify\?[^\s`\"']*\btoken=[A-Za-z0-9._~-]{12,}", re.I)),
    ("secret environment assignment", re.compile(r"\b(?:DATABASE_URL|POSTGRES_PASSWORD|S3_SECRET_ACCESS_KEY|DOKPLOY_API_KEY|AUTH_SECRET)\s*=", re.I)),
    ("URL credentials", re.compile(r"https?://[^\s/:]+:[^\s/@]+@[^\s/]+")),
)
EXCLUDED = {".git", "node_modules", "__pycache__"}
findings: list[str] = []

for path in ROOT.rglob("*"):
    if not path.is_file() or any(part in EXCLUDED for part in path.parts):
        continue
    try:
        content = path.read_text(encoding="utf-8")
    except UnicodeDecodeError:
        continue
    for number, line in enumerate(content.splitlines(), start=1):
        for label, pattern in PATTERNS:
            if pattern.search(line):
                findings.append(f"{path.relative_to(ROOT)}:{number} ({label})")

if findings:
    raise RuntimeError(f"Potential sensitive content found at {', '.join(findings)}")

print("No credentials, browser sessions, bearer JWTs, or live magic links detected in the public source tree.")
